File: dataset/cpp/c05_palindrome_type4.cpp
#include <iostream>
#include <string>
#include <deque>
#include <cctype>
using namespace std;

bool isPalindrome(const string& text) {
    deque<char> dq;

    for (char ch : text) {
        if (isalnum(static_cast<unsigned char>(ch))) {
            dq.push_back(tolower(static_cast<unsigned char>(ch)));
        }
    }

    while (dq.size() > 1) {
        if (dq.front() != dq.back()) {
            return false;
        }
        dq.pop_front();
        dq.pop_back();
    }

    return true;
}

int main() {
    cout << boolalpha << isPalindrome("A man, a plan, a canal: Panama") << endl; // Example usage
    return 0;
}

File: dataset/python_mix/C05_palindrome_type4.py
from collections import deque

def is_palindrome(text: str) -> bool:
    dq = deque(ch.lower() for ch in text if ch.isalnum())
    while len(dq) > 1:
        if dq.popleft() != dq.pop():
            return False
    return True
